httpbakery/discharge.go

```diff
@@ -183,31 +194,41 @@ type dischargeResponse struct {
 
 // Discharge discharges a third party caveat.
 func (h dischargeHandler) Discharge(p httprequest.Params, r *dischargeRequest) (*dischargeResponse, error) {
-	var id []byte
-	if r.Id64 != "" {
-		id1, err := base64.RawURLEncoding.DecodeString(r.Id64)
-		if err != nil {
-			return nil, errgo.Notef(err, "bad base64-encoded caveat id: %v", err)
-		}
-		id = id1
-	} else {
-		id = []byte(r.Id)
+	id, err := maybeBase64Decode(r.Id, r.Id64)
+	if err != nil {
+		return nil, errgo.Notef(err, "bad caveat id")
 	}
 	var caveat []byte
 	if r.Caveat != "" {
+		// Note that it's important that when r.Caveat is empty,
+		// DischargeParams.Caveat is nil.
```

Comment:
It's something we're ensuring with the if above. I've rephrased the comment to hopefully make that clearer.